extension/src/repository/model/collect.ts

```diff
@@ -105,83 +105,122 @@ const addToTracked = (
+    paths: dataStatusOutput.untracked
+  },
+  {
+    key: ExtendedDataStatus.UNCHANGED,
+    paths: dataStatusOutput.unchanged
   }
+]
 
-  for (const [key, data] of createIterable(dataStatus)) {
-    collectStatus(key as ExtendedStatus, data)
+const collectSingleStatus = (
+  acc: { tracked: Set<string> },
+  dataStatusMapping: DataStatusMapping,
+  dvcRoot: string,
+  status: ExtendedStatus,
+  paths: string[] | undefined
+): DataStatusMapping => {
+  for (const path of paths || []) {
+    dataStatusMapping[removeTrailingSlash(path)] = status
+    addToTracked(acc.tracked, resolve(dvcRoot, path), status)
   }
+  return dataStatusMapping
+}
 
-  return { dataStatusMapping, tracked }
+const collectMappingFromGroup = (
+  acc: { tracked: Set<string> },
+  dvcRoot: string,
+  dataStatus: Partial<Record<ExtendedStatus, string[] | undefined>>
+): DataStatusMapping => {
+  const dataStatusMapping: DataStatusMapping = {}
+
+  for (const [key, data] of Object.entries(dataStatus)) {
+    collectSingleStatus(
+      acc,
+      dataStatusMapping,
+      dvcRoot,
+      key as ExtendedStatus,
+      data
+    )
+  }
+
+  return dataStatusMapping
 }
 
-const getInitialDataStatus = (tracked: Set<string>): DataStatus => {
-  const initialDataStatus = {} as DataStatus
+const createDataStatusAccumulator = (): DataStatusAccumulator => {
+  const acc = {} as DataStatusAccumulator
   for (const status of Object.values(AvailableDataStatus)) {
-    initialDataStatus[status] = new Set<string>()
+    acc[status] = new Set<string>()
   }
-  return { ...initialDataStatus, tracked }
+  return acc
 }
 
-export const collectDataStatus = (
+const collectGroupWithMissingAncestors = (
+  acc: DataStatusAccumulator,
   dvcRoot: string,
-  dataStatusOutput: DataStatusOutput & { untracked?: string[] }
-): DataStatus => {
-  const { dataStatusMapping, tracked } = transformDataStatusOutput(
-    dvcRoot,
-    dataStatusOutput
-  )
-
-  const dataStatusWithMissingAncestors =
-    collectMissingAncestors(dataStatusMapping)
-
-  const dataStatus = getInitialDataStatus(tracked)
+  dataStatusMapping: DataStatusMapping
+) => {
+  const groupWithMissingAncestors = collectMissingAncestors(dataStatusMapping)
 
-  for (const [path, status] of Object.entries(dataStatusWithMissingAncestors)) {
+  for (const [path, status] of Object.entries(groupWithMissingAncestors)) {
     const absPath = resolve(dvcRoot, path)
     if (status !== DiscardedStatus.UNCHANGED) {
-      dataStatus[status].add(absPath)
+      acc[status].add(absPath)
     }
 
-    addToTracked(dataStatus.trackedDecorations, absPath, status)
+    addToTracked(acc.trackedDecorations, absPath, status)
   }
+}
 
-  return dataStatus
+export const collectDataStatus = (
+  dvcRoot: string,
+  dataStatusOutput: DataStatusOutput & { untracked?: string[] }
+): DataStatusAccumulator => {
+  const acc = createDataStatusAccumulator()
+
+  for (const groupedStatuses of mapGroupedStatuses(dataStatusOutput)) {
```

Comment:
[F] We now pre-process committed & uncommitted differently than we do unchanged, untracked & not in cache. We do this because paths should be unique in each of those groups but they can be duplicated across groups. After creating an object of the format `{ [path: string]: Status }` for each group we pass it into the old process and fill in missing ancestors. We then transform the data into `{ [status: Status]: Set<string> }` for use in the rest of the subsystem.